File: frontend/src/types/desolvation.ts
/**
 * Desolvation energy calculation types
 * 去溶剂化能计算相关类型定义
 */

/**
 * 去溶剂化模式
 * - stepwise: 逐级去溶剂（每次去掉一个配体）
 * - full: 全部去溶剂（直接计算中心离子单独能量）
 */
export type DesolvationMode = 'stepwise' | 'full';

export interface DesolvationJobCreate {
  md_job_id: number;
  solvation_structure_id: number;
  method_level: string;
  desolvation_mode?: DesolvationMode;  // 默认为 'stepwise'
}

export interface LigandDesolvationResult {
  ligand_id: number;
  ligand_type: string;
  ligand_label: string;
  e_ligand: number;
  e_cluster_minus: number;
  delta_e: number;
}

export interface TypeSummary {
  ligand_type: string;
  avg_delta_e: number;
  std_delta_e: number;
  count: number;
  min_delta_e: number;
  max_delta_e: number;
}

export interface DesolvationEnergyResult {
  id: number;
  postprocess_job_id: number;
  solvation_structure_id: number;
  method_level: string;
  basis_set?: string;
  functional?: string;
  e_cluster: number;
  per_ligand_results: LigandDesolvationResult[];
  per_type_summary: TypeSummary[];
  created_at: string;
}

export interface DesolvationJobResponse {
  job_id: number;
  status: string;
  method_level: string;
  desolvation_mode: DesolvationMode;
  created_at: string;
  started_at?: string;
  finished_at?: string;
  elapsed_seconds?: number;
  error_message?: string;
  result?: DesolvationEnergyResult;
}

File: backend/app/schemas/desolvation.py
"""
Desolvation energy calculation schemas
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class DesolvationJobCreate(BaseModel):
    """创建去溶剂化能任务"""
    md_job_id: int = Field(..., description="MD job ID")
    solvation_structure_id: int = Field(..., description="Solvation structure ID")
    method_level: str = Field(default="standard", description="Calculation method level: fast (6-31G(d)/B3LYP), standard (6-31++G(d,p)/B3LYP), accurate (6-311++G(2d,2p)/wB97XD)")


class LigandDesolvationResult(BaseModel):
    """单个配体的去溶剂化能结果"""
    ligand_id: int = Field(..., description="Ligand ID in the cluster")
    ligand_type: str = Field(..., description="Ligand type (e.g., FSI, EC, DMC)")
    ligand_label: str = Field(..., description="Ligand label (e.g., FSI_1, EC_2)")
    e_ligand: float = Field(..., description="Isolated ligand energy in A.U.")
    e_cluster_minus: float = Field(..., description="Cluster energy without this ligand in A.U.")
    delta_e: float = Field(..., description="Desolvation energy in kcal/mol")


class TypeSummary(BaseModel):
    """按类型汇总的统计"""
    ligand_type: str = Field(..., description="Ligand type")
    avg_delta_e: float = Field(..., description="Average desolvation energy in kcal/mol")
    std_delta_e: float = Field(..., description="Standard deviation in kcal/mol")
    count: int = Field(..., description="Number of ligands of this type")
    min_delta_e: float = Field(..., description="Minimum desolvation energy in kcal/mol")
    max_delta_e: float = Field(..., description="Maximum desolvation energy in kcal/mol")


class DesolvationEnergyResultSchema(BaseModel):
    """去溶剂化能结果"""
    id: int
    postprocess_job_id: int
    solvation_structure_id: int
    method_level: str
    basis_set: Optional[str] = None
    functional: Optional[str] = None
    e_cluster: float = Field(..., description="Complete cluster energy in A.U.")
    per_ligand_results: List[LigandDesolvationResult] = Field(default_factory=list)
    per_type_summary: List[TypeSummary] = Field(default_factory=list)
    created_at: datetime

    class Config:
        from_attributes = True


class DesolvationJobResponse(BaseModel):
    """去溶剂化能任务响应"""
    job_id: int
    status: str
    method_level: str
    created_at: datetime
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    elapsed_seconds: Optional[float] = None
    error_message: Optional[str] = None
    result: Optional[DesolvationEnergyResultSchema] = None

    class Config:
        from_attributes = True

